presto-local-file/src/main/java/com/facebook/presto/localfile/DataLocation.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.localfile;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Longs;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+final class DataLocation
+{
+    private final File location;
+    private final Optional<String> pattern;
+
+    @JsonCreator
+    public DataLocation(
+            @JsonProperty("location") String location,
+            @JsonProperty("pattern") Optional<String> pattern)
+    {
+        requireNonNull(location, "location is null");
+        requireNonNull(pattern, "pattern is null");
+
+        File file = new File(location);
+        checkArgument(file.exists(), "location does not exist");
+        if (pattern.isPresent() && !file.isDirectory()) {
+            throw new IllegalArgumentException("pattern may be specified only if location is a directory");
+        }
+
+        this.location = file;
+        this.pattern = (!pattern.isPresent() && file.isDirectory()) ? Optional.of(".*") : pattern;
```

Comment:
compile the regular expression pattern here so it isn't does for each file scanned